flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,422 @@
+class RowComparator private (
+    /** key positions describe which fields are keys in what order */
+    val keyPositions: Array[Int],
+    /** null-aware comparators for the key fields, in the same order as the key fields */
+    val comparators: Array[NullAwareComparator[_]],
+    /** serializers to deserialize the first n fields for comparison */
+    val serializers: Array[TypeSerializer[_]],
+    /** auxiliary fields for normalized key support */
+    private val auxiliaryFields: (Array[Int], Int, Int, Boolean))
+  extends CompositeTypeComparator[Row] with Serializable {
+
+  // null masks for serialized comparison
+  private val nullMask1 = new Array[Boolean](serializers.length)
+  private val nullMask2 = new Array[Boolean](serializers.length)
+
+  // cache for the deserialized key field objects
+  @transient
+  private lazy val deserializedKeyFields1: Array[Any] = instantiateDeserializationFields()
+
+  @transient
+  private lazy val deserializedKeyFields2: Array[Any] = instantiateDeserializationFields()
+
+  // create auxiliary fields
+  private val normalizedKeyLengths: Array[Int] = auxiliaryFields._1
+  private val numLeadingNormalizableKeys: Int = auxiliaryFields._2
+  private val normalizableKeyPrefixLen: Int = auxiliaryFields._3
+  private val invertNormKey: Boolean = auxiliaryFields._4
+
+  /**
+   * Intermediate constructor for creating auxiliary fields.
+   */
+  def this(
+      keyPositions: Array[Int],
+      comparators: Array[NullAwareComparator[_]],
+      serializers: Array[TypeSerializer[_]]) = {
+    this(
+      keyPositions,
+      comparators,
+      serializers,
+      createAuxiliaryFields(keyPositions, comparators))
+  }
+
+  /**
+   * General constructor for RowComparator.
+   *
+   * @param keyPositions key positions describe which fields are keys in what order
+   * @param comparators non-null-aware comparators for the key fields, in the same order as
+   *   the key fields
+   * @param serializers serializers to deserialize the first n fields for comparison
+   * @param orders sorting orders for the fields
+   */
+  def this(
+      keyPositions: Array[Int],
+      comparators: Array[TypeComparator[_]],
+      serializers: Array[TypeSerializer[_]],
+      orders: Array[Boolean]) = {
+    this(
+      keyPositions,
+      makeNullAware(comparators, orders),
+      serializers)
+  }
+
+  private def instantiateDeserializationFields(): Array[Any] = {
+    val newFields = new Array[Any](serializers.length)
+    var i = 0
+    while (i < serializers.length) {
+      newFields(i) = serializers(i).createInstance()
+      i += 1
+    }
+    newFields
+  }
+
+  // --------------------------------------------------------------------------------------------
+  //  Comparator Methods
+  // --------------------------------------------------------------------------------------------
+
+  override def getFlatComparator(flatComparators: util.List[TypeComparator[_]]): Unit =
+    comparators.map(_.wrappedComparator).foreach {
```

Comment:
The key fields extracted by `extractKeys()` can be null. These keys are compared by the comparators returned by this method. Hence, we must return comparators which are null-aware.